src/qdlutils/applications/qdlscan/main.py

```diff
@@ -0,0 +1,1289 @@
+                                  stop=self.max_position_2, 
+                                  num=n_pixels)
+        # To hold scan results
+        self.data_z = np.empty(shape=(n_pixels, n_pixels))
+        self.data_z[:,:] = np.nan
+
+        # Launch the line scan GUI
+        # Then initialize the GUI
+        self.root = tk.Toplevel()
+        self.root.title(f'Scan {id} ({self.timestamp.strftime("%Y-%m-%d %H:%M:%S")})')
+        self.view = ImageScanApplicationView(window=self.root, 
+                                            application=self,
+                                            settings_dict=parent_application.scan_parameters)
+        
+        # Bind the buttons
+        self.view.control_panel.pause_button.bind("<Button>", self.pause_scan)
+        self.view.control_panel.continue_button.bind("<Button>", self.continue_scan)
+        self.view.control_panel.save_button.bind("<Button>", self.save_scan)
+        self.view.control_panel.norm_button.bind("<Button>", self.set_normalize)
+        self.view.control_panel.autonorm_button.bind("<Button>", self.auto_normalize)
+
+        # Setup the callback for rightclicks on the figure canvas
+        self.view.data_viewport.canvas.mpl_connect('button_press_event', 
+                                                   lambda event: self.open_rclick(event) if event.button == 3 else None)
+        # Set the commands for the right click
+        self.view.rclick_menu.add_command(label='Go to position', command=self.rclick_go_to) 
+        self.view.rclick_menu.add_separator() 
+        self.view.rclick_menu.add_command(label='Open counter', command=self.rclick_open_counter) 
+
+        # Launch the thread
+        self.scan_thread = Thread(target=self.start_scan_thread_function)
+        self.scan_thread.start()
+
+    def continue_scan(self, tkinter_event=None):
+        # Don't do anything if busy
+        if self.application_controller.busy:
+            logger.error('Controller is busy; cannot continue scan.')
+        if self.current_scan_index == self.n_pixels:
+            logger.error('Scan already completed.')
+            return None
+        # Start the scan thread to continue
+        self.scan_thread = Thread(target=self.continue_scan_thread_function)
+        self.scan_thread.start()
+    
+    def pause_scan(self, tkinter_event=None):
+        '''
+        Tell the scanner to pause scanning
+        '''
+        if self.application_controller.stop_scan is True:
+            logger.info('Already waiting to pause.')
+            return None
+        if self.current_scan_index == self.n_pixels:
+            logger.error('Scan already completed.')
+            return None
+
+        if self.application_controller.busy:
+            logger.info('Pausing the scan...')
+            # Query the controller to stop
+            self.application_controller.stop_scan = True
+
+    def save_scan(self, tkinter_event=None):
```

Comment:
make it so that people can change the file type and save to the specified file type

Reply:
To clarify the comment here for others: currently the save file function allows the user to select a file format extension (e.g. ".png") and then save the file. However, if one deletes the extension in the file path input before saving then then the script will not automatically append the ".png" again. The desired behavior is that the file type is chosen based off of what the user selects from the dropdown box.

Unfortunately this is not something that is possible with Tkinter (https://stackoverflow.com/questions/45823473/tkinter-return-file-type-from-save-dialogue) the file type selection is merely a filter for the results that are shown to the user in the dialogue box. You cannot get which one was chosen by the user.

The desired behavior could be achieved by disabling any choice the user makes about file type and saving the png and hdf5 in both cases. But this is undesirable since you may not always want to save the image itself.